### PR TITLE
Autocomplete JavaScript: Part 2

### PR DESCRIPTION
**Pull request by mmurray22**
In this continuation of the previous autocomplete part one, we make the final connections between the front-end JavaScript and the servlet which will be fetching the actual information we wish to display to the user.

Notes on resiliency: Currently resilient to weird casing (upper case, lower case, mix) and trailing/leading spaces. Cannot add a college to the name queue if it's already there. Also, identifies the correct location in the string to bold (for most cases). There is one remaining bug present will be addressed in another PR, as described below.

Remaining bug: The scorecard api suggests schools like 'St. John's University-New York', for input like 'New York U', so that messes with the frontend. Will be addressed in a subsequent PR since this involves backend changes.

### CODE REVIEW

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java [hunk 1]
```diff
@@ -38,4 +40,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    List<String> listToSend = new ArrayList<>();
+    for (int i = 0; i < numberOfColleges; i++) {
+      StringBuilder parameterStr = new StringBuilder();
+      parameterStr.append("college");
+      parameterStr.append(Integer.toString(i));
+      String collegeName = request.getParameter(parameterStr.toString());
+      listToSend.add(collegeName);
+    }
+    RequestDispatcher rd = request.getRequestDispatcher("map");
+    rd.forward(request, response);
```
Comment (ihsan314) [inline capstone/src/main/java/com/google/univiz/servlets/SearchService.java:57]: Are you only forwarding to the map servlet?

File: capstone/src/main/webapp/js/servletConnection.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Gets suggestions to display to the users
+ * @return{void}
```
Comment (ihsan314) [inline capstone/src/main/webapp/js/servletConnection.js:4]:
```suggestion
```

File: capstone/src/main/webapp/js/servletConnection.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ */
+function getSuggestions() {
+  fetch('/search').then((response) => response.json()).then((suggestions) => {
+    provideSuggestions(suggestions);
+  });
+}
+
+/**
+ * Sends final college list to servlet
+ * @return{void}
```
Comment (ihsan314) [inline capstone/src/main/webapp/js/servletConnection.js:14]:
```suggestion
```
No need to document the lack of a return type.

File: capstone/src/main/webapp/js/servletConnection.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Gets suggestions to display to the users
+ * @return{void}
+ */
+function getSuggestions() {
+  fetch('/search').then((response) => response.json()).then((suggestions) => {
+    provideSuggestions(suggestions);
```
Comment (ihsan314) [inline capstone/src/main/webapp/js/servletConnection.js:8]:
```suggestion
  fetch('/search').then((response) => response.json())
      .then(provideSuggestions);
```
Since it's expecting a function object, you might as well provide the function directly instead of wrapping it in an arrow function.

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java [hunk 5]
```diff
@@ -38,4 +40,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    String numberCollegesStr = request.getParameter("collegeNumber");
+    int numberOfColleges = Integer.parseInt(numberCollegesStr);
+    List<String> listToSend = new ArrayList<>();
+    for (int i = 0; i < numberOfColleges; i++) {
+      StringBuilder parameterStr = new StringBuilder();
+      parameterStr.append("college");
+      parameterStr.append(Integer.toString(i));
+      String collegeName = request.getParameter(parameterStr.toString());
+      listToSend.add(collegeName);
+    }
```
Comment (ihsan314) [inline capstone/src/main/java/com/google/univiz/servlets/SearchService.java:55]: What happens to the list after it's been created?

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java [hunk 6]
```diff
@@ -38,4 +40,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String numberCollegesStr = request.getParameter("collegeNumber");
+    int numberOfColleges = Integer.parseInt(numberCollegesStr);
+    List<String> listToSend = new ArrayList<>();
+    for (int i = 0; i < numberOfColleges; i++) {
+      StringBuilder parameterStr = new StringBuilder();
+      parameterStr.append("college");
+      parameterStr.append(Integer.toString(i));
```
Comment (ihsan314) [inline capstone/src/main/java/com/google/univiz/servlets/SearchService.java:52]: Should there be a separator, like a colon or comma?

File: capstone/src/main/webapp/js/servletConnection.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * Sends final college list to servlet
+ * @return{void}
+ */
+function sendCollegeInformation() {
+  const params = new URLSearchParams();
+  const listItems = document.querySelector('ul').children;
+  const listArray = Array.from(listItems);
+  const index = 0;
+  for (elt in listArray) {
```
Comment (biancamacias) [inline capstone/src/main/webapp/js/servletConnection.js:21]:
```suggestion
  for (const elt of listArray) {
```
per [style guide](https://google.github.io/styleguide/jsguide.html#features-for-loops)

File: capstone/src/main/webapp/js/servletConnection.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/**
```
Comment (biancamacias) [inline capstone/src/main/webapp/js/servletConnection.js:1]: you will probably have to add an `no-unused-vars` error ignore for some of these functions as they might be getting called from elsewhere. look at #100 or #35 for reference.

File: capstone/src/main/webapp/index.html [hunk 9]
```diff
@@ -8,13 +8,14 @@
   </head>
   <body>
     <h2>Welcome to UniViz!</h2>
-    <form autocomplete="off">
+    <!-- TODO: This form needs the most work (need to get data async) -->
+    <form action="/search" autocomplete="off" type="GET" id="search-bar">
       <label for="search">Search for colleges:</label><br>
-      <input type="text" id="search" name="search" placeholder="College Name" oninput="giveSuggestions()" onkeydown="keyDown(event)"><br>
-      <input type="submit" value="See your colleges!">
+      <input type="text" id="search" name="query" placeholder="College Name" oninput="giveSuggestions()" onkeydown="keyDown(event)"><br>
```
Comment (biancamacias) [inline capstone/src/main/webapp/index.html:14]: im not seeing this function in your js files? will it be implemented later or is there a naming inconsistency?

File: capstone/src/main/webapp/index.html [hunk 10]
```diff
@@ -8,13 +8,14 @@
...
     </form>
     <p>Current List of Suggested Colleges:</p>
     <ul id="suggestions"></ul>
+    <button id="submit" onclick="sendCollegeInformation()">See your colleges' information!</button>
```
Comment (biancamacias) [inline capstone/src/main/webapp/index.html:18]: this function too, i dont see

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java (hunk 6 above)
Comment (mmurray22) [reply capstone/src/main/java/com/google/univiz/servlets/SearchService.java:52]: for the parameter name? no-I made the parameters register themselves as 'college0', 'college1', etc.

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java (hunk 5 above)
Comment (mmurray22) [reply capstone/src/main/java/com/google/univiz/servlets/SearchService.java:55]: It needs to be forwarded to CollegeVis and Maps servlets (trying to figure that out now haha)

File: capstone/src/main/webapp/index.html (hunk 9 above)
Comment (mmurray22) [reply capstone/src/main/webapp/index.html:14]: I am trying to determine how to fit the async call to the API in here, and in doing say I've shuffled quite a bit around...I'd say it's probably a naming inconsistency

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java (hunk 6 above)
Comment (ihsan314) [reply capstone/src/main/java/com/google/univiz/servlets/SearchService.java:52]: Oh I see, that makes sense then

File: capstone/src/main/webapp/index.html (hunk 10 above)
Comment (mmurray22) [reply capstone/src/main/webapp/index.html:18]: It's in the js file :)

File: capstone/src/main/webapp/index.html (hunk 9 above)
Comment (mmurray22) [reply capstone/src/main/webapp/index.html:14]: also looks like giveSuggestions is also in my java file :)

File: capstone/src/main/webapp/js/autocomplete.js [hunk 11]
```diff
@@ -12,6 +12,34 @@ const DOWN_KEY = 'ArrowDown';
...
+  console.log('FetchStr: ', fetchStr);
+  fetch(fetchStr).then((response) => response.json()).then((suggestions) => {
+    console.log('Here! ', suggestions);
+    giveSuggestions(suggestions);
+  });
+}
+
+/**
+ * Sends final college list to the dashboard
+ * @return{void}
```
Comment (ihsan314) [inline capstone/src/main/webapp/js/autocomplete.js:30]:
```suggestion
```

File: capstone/src/main/webapp/js/autocomplete.js [hunk 12]
```diff
@@ -115,8 +118,9 @@ function displaySuggestions(trimArr, autocompleteList, val) {
 
 /**
  * Event occurrance when input is provided to autocomplete field
+ * @param{Object} suggestions
```
Comment (ihsan314) [inline capstone/src/main/webapp/js/autocomplete.js:121]: If you can find a more specific type than `Object` that would be nice. If not, then you can describe `suggestions` in a way that makes it's format clear. As it stands I'm not getting a clear idea of what `suggestions` is supposed to look like.

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java [hunk 13]
```diff
@@ -17,7 +17,8 @@
  * Servlet that returns the list of matching college names based on a partial input query. This
  * service is best used for the auto-complete feature.
  */
-@WebServlet("/search")
+// @WebServlet("/search")
```
Comment (kaveh096) [inline capstone/src/main/java/com/google/univiz/servlets/SearchService.java:20]: nit: remove this comment

File: capstone/src/main/webapp/js/autocomplete.js [hunk 14]
```diff
@@ -12,6 +12,31 @@ const DOWN_KEY = 'ArrowDown';
 const ENTER_KEY = 'Enter';
 let selectedSuggestionPosition = -1;
 
+/**
+ * Gets suggestions to display to the users
+ */
+function getSuggestions() {
+  const fetchStr = '/search?query=' + SEARCH_INPUT.value;
+  fetch(fetchStr).then((response) => response.json()).then((suggestions) => {
+    giveSuggestions(suggestions);
```
Comment (kaveh096) [inline capstone/src/main/webapp/js/autocomplete.js:21]: same sugegstion about method naming [in part 1](https://github.com/googleinterns/univiz/pull/50/files/1e428a8aacd5b996e3fe6fcb87150648e123ba1b#r478783432).

File: capstone/src/main/webapp/js/autocomplete.js [hunk 15]
```diff
@@ -12,6 +12,31 @@ const DOWN_KEY = 'ArrowDown';
...
+ */
+function sendCollegeInformation() {
+  const listItems = document.querySelector('ul').children;
+  const listArray = Array.from(listItems);
+  let urlHtml = 'dashboard.html?id=';
+  for (const elt of listArray) {
+    urlHtml += elt.id;
+    urlHtml += ',';
+  }
+  urlHtml = urlHtml.slice(0, -1);
```
Comment (kaveh096) [inline capstone/src/main/webapp/js/autocomplete.js:36]: if the id list was empty, this slice will end up deleting the `=` form the url. So you'll need another if here.

Alternatively you can use Array methods:

```js
idArray = Array.from(listArray, l => l.id)
let urlHtml = 'dashboard.html?id=' + idArray.join();
```

File: capstone/src/main/webapp/js/autocomplete.js [hunk 16]
```diff
@@ -12,6 +12,31 @@ const DOWN_KEY = 'ArrowDown';
...
+  window.location.href = urlHtml;
```
Comment (kaveh096) [inline capstone/src/main/webapp/js/autocomplete.js:37]: Should we redirect to dashboard if listArray is empty? No action items for this PR. Something to think about.

File: capstone/src/main/java/com/google/univiz/servlets/SearchService.java (hunk 13 above)
Comment (mmurray22) [reply capstone/src/main/java/com/google/univiz/servlets/SearchService.java]: done!

File: capstone/src/main/webapp/js/autocomplete.js (hunk 15 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js:36]: wow didn't know that existed! I'll use that :)

File: capstone/src/main/webapp/js/autocomplete.js (hunk 14 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js:21]: alrighty, I'll go through and try to make the naming scheme more sensible! thanks!

File: capstone/src/main/webapp/js/autocomplete.js [hunk 17]
```diff
@@ -12,6 +12,32 @@ const DOWN_KEY = 'ArrowDown';
...
+  fetch(fetchStr).then((response) => response.json()).then((suggestions) => {
+    getsAllProposedSuggestions(suggestions, val);
+  });
+}
+
+/**
+ * Sends final college list to the dashboard
+ */
+function sendCollegeInformationToDashboard() {
+  const listItems = document.querySelector('ul').children;
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:34]: This feels a bit brittle. What if another `ul` appears in the DOM before this `ul`? Can we refer to the ul by an id here, instead?

File: capstone/src/main/webapp/js/autocomplete.js [hunk 18]
```diff
@@ -12,6 +12,32 @@ const DOWN_KEY = 'ArrowDown';
...
+  const listArray = Array.from(listItems);
+  const idArray = Array.from(listArray, (l) => l.id);
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:36]: nit - I think this might be more intuitive with a `map` invocation.

```suggestion
  const idArray = Array.from(listItems).map((l) => l.id).join();
```

File: capstone/src/main/webapp/js/autocomplete.js [hunk 19]
```diff
@@ -12,6 +12,32 @@ const DOWN_KEY = 'ArrowDown';
...
+  const urlHtml = 'dashboard.html?id=' + idArray.join();
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:37]: what does the `Html` in `urlHtml` refer to? Would something like `dashboardUrl` be more descriptive?

File: capstone/src/main/webapp/js/autocomplete.js [hunk 20]
```diff
@@ -57,56 +83,45 @@ function closeAllElmntExcept(elmnt) {
...
- */
-function getListOfSuggestions() {
-  const arr = ['Hallo', 'Hello', 'Hi', 'Hiya', 'Howdy', 'Wassup'];
-  return arr;
-}
-
-/**
- * Identifies and returns relevant suggestions in the arr
- * @param {string[]} arr
+ * Creaters List Element to display a college name suggestion
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:95]:
```suggestion
 * Creates List Element to display a college name suggestion
```

File: capstone/src/main/webapp/js/autocomplete.js [hunk 21]
```diff
@@ -57,56 +83,45 @@ function closeAllElmntExcept(elmnt) {
...
-  return relevantSuggestions;
+function createListElmt(arrElt, val) {
+  const listElmt = document.createElement('div');
+  const collegeNameUpper = arrElt.collegeName.toUpperCase();
+  const valIndex = collegeNameUpper.indexOf(val.toUpperCase());
+  listElmt.innerHTML = arrElt.collegeName.substr(0, valIndex);
+  listElmt.innerHTML += '<strong>' +
+                       arrElt.collegeName.substr(valIndex, val.length) +
+                       '</strong>';
+  listElmt.innerHTML += arrElt.collegeName.substr(val.length + valIndex);
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:108]: This seems like it would benefit from some unit testing :)

File: capstone/src/main/webapp/index.html [hunk 22]
```diff
@@ -7,13 +7,11 @@
   </head>
   <body>
     <h2>Welcome to UniViz!</h2>
-    <form autocomplete="off">
-      <label for="search">Search for colleges:</label><br>
-      <input type="text" id="search" name="search" placeholder="College Name" oninput="giveSuggestions()" onkeydown="keyDown(event)"><br>
-      <input type="submit" value="See your colleges!">
-    </form>
+    <label for="search">Search for colleges:</label><br>
```
[the remark this answers is not in the document]
Comment (mmurray22) [reply capstone/src/main/webapp/index.html:10]: the form element is not needed for the autocomplete functionality (it was never actually used-I thought it would be, but turns out I didn't need it!)

File: capstone/src/test/javascript/search/autocompleteSpec.js [hunk 23]
```diff
@@ -1,15 +0,0 @@
-describe('Autocomplete Suggests correct array', () => {
```
[the remark this answers is not in the document]
Comment (mmurray22) [reply capstone/src/test/javascript/search/autocompleteSpec.js:1]: yes will do! I'll add back the tests-especially since there are definitely some new functions to test!

File: capstone/src/main/webapp/js/autocomplete.js (hunk 19 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js]: great point! changed!

File: capstone/src/main/webapp/js/autocomplete.js (hunk 21 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js:108]: done!

File: capstone/src/main/webapp/js/autocomplete.js (hunk 17 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js:34]: updated :)

File: capstone/src/main/webapp/js/autocomplete.js [hunk 24]
```diff
@@ -12,6 +12,31 @@ const DOWN_KEY = 'ArrowDown';
...
+    getsAllProposedSuggestions(suggestions, val);
+  });
+}
+
+/**
+ * Sends final college list to the dashboard
+ */
+function sendCollegeInformationToDashboard() {
+  const listItems = document.getElementById('suggestions').children;
+  const listArray = Array.from(listItems);
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:34]: `listArray` can be removed now.

File: capstone/src/main/webapp/js/autocomplete.js [hunk 25]
```diff
@@ -12,6 +12,31 @@ const DOWN_KEY = 'ArrowDown';
...
+  const idArray = Array.from(listItems).map((l) => l.id).join(',');
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:35]: nit - consider renaming this variable since it's no longer an array.

File: capstone/src/main/webapp/js/autocomplete.js [hunk 26]
```diff
@@ -57,56 +82,45 @@ function closeAllElmntExcept(elmnt) {
...
-function getRelevantDataSuggestions(arr, val) {
-  const relevantSuggestions = [];
-  val = val.toUpperCase();
-  for (const arrElt of arr) {
-    if (arrElt.substr(0, val.length).toUpperCase() === val.toUpperCase()) {
-      relevantSuggestions.push(arrElt);
-    }
-  }
-  return relevantSuggestions;
+function createListElmt(collegeName, val) {
```
Comment (ccondit) [inline capstone/src/main/webapp/js/autocomplete.js:99]: this function name is very generic. can you refine it?

File: capstone/src/test/javascript/search/autocompleteSpec.js [hunk 27]
```diff
@@ -1,15 +1,16 @@
...
-  it('Autocomplete selects correct item', () => {
-    const fullData = ['Ahoy', 'Hello', 'Wattup'];
-    const relevantData = ['Hello'];
-    const retData = getRelevantDataSuggestions(fullData, 'H');
-    expect(retData).toEqual(relevantData);
+describe('Autocomplete provides correct list element', () => {
+  it('Tests Proper List Element Creation Simple', () => {
+    const arrElt = 'Hello';
```
Comment (ccondit) [inline capstone/src/test/javascript/search/autocompleteSpec.js:3]: can you give these variables more descriptive names? `arrElt` and `val` don't give the reader much to go on...

File: capstone/src/test/javascript/search/autocompleteSpec.js [hunk 28]
```diff
@@ -1,15 +1,16 @@
...
+    const listElmtExpected = '<strong>He</strong>llo';
+    expect(listElmtExpected).toEqual(listElmtActual.innerHTML);
   });
-
-  it('Autocomplete selects multiple correct items', () => {
-    const fullData = ['Ahoy', 'Hello', 'Hi', 'Wattup'];
-    const relevantData = ['Hello', 'Hi'];
-    const retData = getRelevantDataSuggestions(fullData, 'H');
-    expect(retData).toEqual(relevantData);
+  it('Tests Proper List Element Creation Complex', () => {
```
Comment (ccondit) [inline capstone/src/test/javascript/search/autocompleteSpec.js:9]: How about adding a test for a corner case where the completion is at the end?

File: capstone/src/main/webapp/js/autocomplete.js (hunk 24 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js]: done!

File: capstone/src/main/webapp/js/autocomplete.js (hunk 25 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js]: done!

File: capstone/src/main/webapp/js/autocomplete.js (hunk 26 above)
Comment (mmurray22) [reply capstone/src/main/webapp/js/autocomplete.js]: gave it a new name-I hope it's more clarified!

File: capstone/src/test/javascript/search/autocompleteSpec.js (hunk 27 above)
Comment (mmurray22) [reply capstone/src/test/javascript/search/autocompleteSpec.js]: updated!

File: capstone/src/test/javascript/search/autocompleteSpec.js (hunk 28 above)
Comment (mmurray22) [reply capstone/src/test/javascript/search/autocompleteSpec.js]: added!
